[DA] runtime predicates for delinearization bounds checks

sebpop opened the pull request:
When compile-time checks fail, rely on runtime SCEV predicates, instead of failing delinearization entirely.  This allows delinearization to succeed in more cases where compile-time proofs are not possible, enabling more precise dependence analysis under runtime assumptions.

sebpop committed:
[DA] runtime predicates for delinearization bounds checks

When compile-time checks fail, rely on runtime SCEV predicates, instead of
failing delinearization entirely.  This allows delinearization to succeed in
more cases where compile-time proofs are not possible, enabling more precise
dependence analysis under runtime assumptions.

sebpop committed:
[DA] runtime predicates for overflow checks in delinearization

When compile-time overflow checks (for Prod, Min, and Max offset
computations) fail, add runtime SCEV predicates using the equality-based
overflow detection pattern: (sext A) op (sext B) == sext(A op B). This
allows delinearization to succeed in more cases where compile-time proofs
are not possible, enabling more precise dependence analysis under runtime
assumptions.

This extends the runtime predicate support from PR #170713 to also
cover the overflow validation checks added in PR #169902.

kasuga-fj reviewed (commented):
High level thought: I think the runtime prediction feature should be removed for now, and re-introduced after the functionality like `BatchDA` is implemented. Current implementation looks a bit ugly to me, since each time the `depends` is invoked, same predicates will keep getting added. Probably a better design would be to use a single set of assumptions throughout the analysis.

sebpop committed:
[DA] Add getNoOverflowPredicate() to ScalarEvolution for reuse

Refactor the overflow check runtime predicate generation to reuse the
same pattern used by ScalarEvolution::willNotOverflow(). Instead of
duplicating the ext(LHS op RHS) == ext(LHS) op ext(RHS) pattern in
Delinearization.cpp, add a new getNoOverflowPredicate() method to
ScalarEvolution that returns the predicate (or nullptr if no-overflow
is already provable at compile time).

This addresses review feedback to avoid code duplication between
willNotOverflow() and the runtime predicate generation in delinearization.

sebpop committed:
[DA] Deduplicate runtime predicates in delinearization

When validating both source and destination subscripts, the same
predicate can be generated multiple times (e.g., when both access
A[i][j]). Add a helper AddPredicate() that checks if a predicate
already exists before adding it to avoid duplicates.

nikic commented:
Just as a high-level question: Do we want to introduce more runtime predicates at this point in time? Given that there is a lot of active ongoing work on correctness issues in DA/Delinearization, it may be better to delay this until a later time, just to make it easier to do further code changes, without also having to maintain the predicates along the way.

I've been out for a while and am not up to date on the DA/Delinearization work, but IIRC these runtime predicates are not yet used anywhere, right? They only appear in the analysis output for now? I expect that actually making use of them will be quite challenging, because of the need to trade off the value of the transform with the probability of passing the runtime check and the code size increase of performing loop versioning. Having a lot of runtime checks implemented will probably make this harder than introducing them gradually and evaluating which of them are actually valuable in practice.

On the other hand, I could see value in doing this now if this helps with testing somehow, e.g. because it allows us to easily test situations in DA that would otherwise be hard to test because of delinearization failures. Is something like this the case?
